flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/resource/ResourceManager.java

```diff
@@ -419,4 +527,26 @@ private void registerResources(
                     LOG.info("Register resource [{}] successfully.", resourceUri.getUri());
                 });
     }
+
+    /**
+     * Resource with reference counter, when the counter is 0, it means the resource can be removed.
+     */
+    static class ResourceCounter {
+        final URL url;
+        int counter;
+
+        private ResourceCounter(URL url) {
+            this.url = url;
+            this.counter = 0;
+        }
+
+        private void increaseCounter() {
+            this.counter++;
+        }
+
+        private boolean decreaseCounter() {
+            this.counter--;
+            return this.counter == 0;
```

Comment:
We'd better add a safetynet here in case the counter incorrectly decreased to negative.